app/crud.py

```diff
@@ -0,0 +1,87 @@
+from typing import Sequence, Dict
+
+import httpx
+from sqlalchemy.ext.asyncio import AsyncSession
+from sqlalchemy.future import select
+
+from app import models, schemas
+from app.api import weather
+
+
+class Service:
+    def __init__(self, session: AsyncSession):
+        self.session = session
+
+    async def get_cities(self) -> Sequence[models.DBCity]:
+        query = select(models.DBCity)
+        result = await self.session.execute(query)
+        return result.scalars().all()
+
+    async def get_city(self, city_id: int) -> models.DBCity:
+        query = select(models.DBCity).where(
+            models.DBCity.id == city_id
+        )
+        result = await self.session.execute(query)
+
+        return result.scalars().first()
+
+    async def create_city(
+            self, city: schemas.CityCreate
+    ) -> models.DBCity:
+        created_city = models.DBCity(
+            name=city.name,
+            additional_info=city.additional_info
+        )
+        self.session.add(created_city)
+        await self.session.commit()
+        await self.session.refresh(created_city)
+        return created_city
+
+    async def update_city(
+        self, city_id: int, city: schemas.CityCreate
+    ) -> models.DBCity:
+        updated_city = await self.get_city(city_id)
+
+        if updated_city:
+            updated_city.name = city.name
+            updated_city.additional_info = city.additional_info
+            await self.session.commit()
+            await self.session.refresh(updated_city)
+        return updated_city
+
+    async def delete_city(self, city_id: int) -> models.DBCity:
+        city = await self.get_city(city_id)
+
+        if city:
+            await self.session.delete(city)
+            await self.session.commit()
+        return city
+
+    async def get_temperatures(
+            self, city_id: int = None
+    ) -> Sequence[models.DBTemperature]:
+        query = select(models.DBTemperature)
+
+        if city_id:
+            query = query.where(models.DBTemperature.city_id == city_id)
+
+        result = await self.session.execute(query)
+        return result.scalars().all()
+
+    async def create_temperatures(
+            self
+    ) -> Dict:
+        cities = await self.get_cities()
+
+        async with httpx.AsyncClient() as client:
+            for city in cities:
+                temperature = await weather.fetch_temperatures(client, city)
+                temperature = models.DBTemperature(**temperature.dict())
+
+                self.session.add(temperature)
+                await self.session.commit()
+                await self.session.refresh(temperature)
+
+        return {
+            "detail": "Temperatures successfully updated"
+        }
```

Comment:
The problem with this approach:

1. Tight Coupling with ORM Models
The Service class is tightly coupled with the ORM models (models.DBCity, models.DBTemperature). This makes it difficult to test the service in isolation because it directly depends on the database models.
Improvement: Use Data Transfer Objects (DTOs) or schemas for input and output. This way, the service methods can return plain Python objects or schemas, which are easier to mock and test.

2. Lack of Transaction Management
The create_city, update_city, and delete_city methods commit the session after each operation. If you need to perform multiple operations within a single transaction, this approach could lead to issues.

3. Error Handling
There's no error handling in any of the methods. If something goes wrong (e.g., the city doesn't exist, a database constraint is violated, etc.), it could cause unhandled exceptions.

4. Session Management in create_temperatures
The create_temperatures method makes multiple commit() calls within a loop, which could lead to partial updates if something fails in the middle.

5. Inefficient Use of ORM
The current approach can lead to inefficient queries, especially in large datasets. For example, fetching all cities or temperatures at once might not scale well.

6. Violation of Single Responsibility Principle (SRP)
The Service class handles both business logic and direct interactions with the database. This violates the SRP, making the class harder to maintain and extend.